# Technical Challenge ! Custiomers and Orders - The lifeblood of any business!

This notebook depends on two data files: customers.scv and orders.csv. These files are hosted in the cloud (or Github), so for simplicity, we are providing you with the code to download and save the two files in a folder called 'sample_data' which lives within this notebook session (not on your local computer nor Google Drive).

![sample_data folder in notebook session](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/sample_data_folder.png)

Please execute the first code snippet before moving to the rest of the exercise as this code will import the data from the files for you to be able to execute the rest of the exercise.

Remember as we said in the previous lecture, you have to save this notebook into your Google Drive by going to menu File/Save a Copy In Drive, or clicking in the button "Copy to Drive". This way, you will not lose any work you've done and the file will persist with your latest changes in your Google Drive. Also we recommend you to rename your .ipynb in your Google Drive, so you can easily find it in the future.

In [ ]:
import requests

def import_data_files():
  r = requests.get('https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/customers.csv')
  with open('./sample_data/customers.csv', 'wb') as f:
    f.write(r.content)

  r = requests.get('https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/orders.csv')
  with open('./sample_data/orders.csv', 'wb') as f:
    f.write(r.content)

import_data_files()
print("Customers and orders CSV files have been added './sample_data'")

Customers and orders CSV files have been added './sample_data'


# Exercise 1: Processing Customers data (difficulty medium)

The sample customer data in 'customers.csv' file has just 5 columns: CustomerId, First Name, Last Name, City and State

![Data sample](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/customers.png)

We strongly recommend that you complete from the [Prep Course: Intro to Python](https://colab.research.google.com/github/anyoneai/notebooks/blob/main/python3_crash_course.ipynb) the following section:
- Section 7: File I/O (to understand how to read CSV file)
- Section 6: For Loop (to navigate the contents of the CSV file)
- Section 5: Tuples, Lists, and Dictionaries (to manipulate the data of the CSV file)

With this, we hope you can complete this exercise successfully. Although if you want to solve this with libraries or any other way, you are welcome to do it your way.

*Hint:* We advise you to take a look at the data before you start.
**if you want to manually take a look at the data before you start, please see the content of the data [here](https://github.com/anyoneai/notebooks/blob/main/customers_and_orders/data/customers.csv).

*Hint*: There are many ways to do this exercise you can do your own, although here's some help. You can solve this exercise reading and parsing CSV files, structuring data into dictionaries, and using for loops to navigate the contents

*Hint*: Also, keep in mind that data might not be clean and you might have to figure out how to deal with that data from the code, without having to modify the data source.

Below are the 5 questions you'll have to answer to pass the evaluation:

**Question 1:** How many customers are in the file?
(as help, we have added some comments and starter code to help you structure the solution)

In [ ]:
from os.path import exists
import csv

# check if file can be found and open the file
datafile = "./sample_data/customers.csv"
if not exists(datafile):
  raise SystemExit("You should run the first code cell and download the dataset files!")

# customers file is in the sample_data folder and can be opened.
# Using csv reader is one way of dealing with the file
# but it's up to you how you want to proceed
with open(datafile, 'r') as fl:
  csvreader = csv.reader(fl, delimiter=',')

  # loop through the rows and count etc. THE REST IS UP TO YOU!


print("All done!")

All done!


In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv("./sample_data/customers.csv")
df.head()

CustomerID    FirstName      LastName         City State
0        6192        Randi    Piedrahita  Canoga Park    CA
1     5100595  Christopher       Abraham   Loma Linda    CA
2     1902451        Elvis      Jauregui  Los Angeles    CA
3     1906377         Jose  Mota-Sanchez       Downey    CA
4     1902406      Scott        McCrorie   Long Beach    CA

In [ ]:
df['State'].unique()

array(['CA', 'AZ', 'NV', 'FL', 'CA           ', 'WA', 'NH', 'ID', 'CO',
       'Ca', 'TX', 'NM', 'OR', 'UT', 'MA', 'IN'], dtype=object)

In [ ]:
df['State'] = df['State'].str.upper()
df['State'] = df['State'].str.strip()

In [ ]:
print('Ther are ', df['CustomerID'].nunique(), ' Customers in the file ', 'and ', df['CustomerID'].count(), 'records')

Ther are  471  Customers in the file  and  602 records


**Question 2:** In how many different states do the customers live in?

In [ ]:
df['State'].unique()

array(['CA', 'AZ', 'NV', 'FL', 'WA', 'NH', 'ID', 'CO', 'TX', 'NM', 'OR',
       'UT', 'MA', 'IN'], dtype=object)

In [ ]:
print('The customers live in ', df['State'].nunique(), ' Different States')

The customers live in  14  Different States


**Question 3** What is the state with most customers?

In [ ]:
df['State'].value_counts()


CA    569
NV      8
AZ      6
FL      3
CO      3
NM      3
TX      2
UT      2
WA      1
NH      1
ID      1
OR      1
MA      1
IN      1
Name: State, dtype: int64

The State with more customers is CA with 569

**Question 4** What is the state with the least customers?

The States with the least customers are: ID, IN, MA, NH, OR, WA

In [ ]:
# open file and init vars
# evaluate the contents of the file

**Question 5:** What is the most common last name?

In [ ]:
df['LastName'].value_counts()

Smith       8
Zambrana    5
Gomez       5
Doggett     5
Rocha       4
           ..
Bagale      1
Puno        1
Kovelman    1
Vrbka       1
Okonma      1
Name: LastName, Length: 424, dtype: int64

The most common last name is Smith

# Exercise 2: Processing Orders data (difficulty high)

The second sample files contains orders placed by customers from the first file. Be careful, this file has many rows and you most likely should not print the contents of the whole file.

The file contains the following columns: CustomerID, OrderID, Date, OrderTotal, ProductName, Price

![Data sample](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/orders.png)

*Hint:* We advise you to take a look at the data before you start.
**if you want to manually take a look at the data before you start, please see the content of the data [here](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/orders.csv).

*Hint*: There are many ways to do this exercise you can do your own, although here's some help. You can solve this exercise reading and parsing CSV files, structuring data into dictionaries, and using for loops to navigate the contents

*Hint*: Also, the data is not clean and you will have to figure out how to deal with that data from the code, without having to modify the data source.



**Question #1:** How many unique orders are in the orders.csv file?

**Question #2:** What is the average number of items per order (rounded to two decimal places)?

**Question #3:** What is the highest number of items per order?

**Question #4:** What is the number of orders placed in October 2021?

**Question #5:** Which customer spent the most amount of money in 2021?

**Question #6:** Historically, what is the best month for sales?

Once you get your answers, remember to go back to the course and introduce them in the multiple choice quiz

In [60]:
from os.path import exists
import csv

# check if file can be found and open the file
datafile = "./sample_data/orders.csv"
if not exists(datafile):
  raise SystemExit("You should run the first code cell and download the dataset files!")

# customers file is in the sample_data folder and can be opened.
# Using csv reader is one way of dealing with the file
# but it's up to you how you want to proceed
with open(datafile, 'r') as fl:
  csvreader = csv.reader(fl, delimiter=',')

  # loop through the rows and count etc. THE REST IS UP TO YOU!


print("All done!")

All done!


In [61]:
df = pd.read_csv("./sample_data/orders.csv")
df.head()

CustomerID  OrderID                     Date  OrderTotal ProductName  Price
0        8091  7742581  2021-07-26 14:40:10.783        95.0         Z03   90.0
1      902139  7742778  2021-08-08 05:01:21.120        60.0    0844 A/C   60.0
2     2300266  7742593  2021-07-27 11:00:16.020       185.0         M07   90.0
3     2300266  7742593  2021-07-27 11:00:16.020       185.0         M09   90.0
4     5173013  7742609  2021-07-28 14:26:13.930       165.0        0324  160.0

In [62]:
# Question #1: How many unique orders are in the orders.csv file?

print('There are ', df['OrderID'].nunique(), 'unique orders in the file')

There are  16672 unique orders in the file


In [63]:
# Question #2: What is the average number of items per order (rounded to two decimal places)?
average = df.groupby('OrderID').count()
print('the average number of ithems per order is: ',round(average['OrderTotal'].mean(),2))

the average number of ithems per order is:  1.76


In [64]:
# Question #3: What is the highest number of items per order?
print('The highest number of items pero orden is: ', average['OrderTotal'].max())

The highest number of items pero orden is:  35


In [72]:
# Question #4: What is the number of orders placed in October 2021?
df['Date'] = pd.to_datetime(df['Date'])
october21 = df.query('Date >= "2021-10-01" & Date <= "2021-10-31"')
print('The numer of orders placed in October 21 is: ',october21['OrderID'].count())

The numer of orders placed in October 21 is:  437


In [80]:
# Question #5: Which customer spent the most amount of money in 2021?
year21 = df.query('Date.dt.year == 2021')
year21.groupby('CustomerID')['Price'].sum().reset_index(name='total_spent').sort_values(by='total_spent', ascending=False)


CustomerID  total_spent
186     5172443       7675.0
118     1907160       5485.0
13         5464       5470.0
172     5100984       5225.0
139     4300438       4960.0
..          ...          ...
290     5370602        100.0
327     5415211        100.0
333     5415286        100.0
278     5370422         90.0
275     5370388         85.0

[342 rows x 2 columns]

The Customer who spent the most amount of money in 2021 was 5172443

In [83]:
# Question #6: Historically, what is the best month for sales?
df.groupby(df['Date'].dt.strftime('%m'))['OrderTotal'].sum().reset_index(name='Total').sort_values(by='Total')

Date     Total
11   12  510948.0
2    03  556715.0
3    04  560733.0
7    08  575493.0
6    07  585495.0
9    10  597649.0
10   11  745817.0
8    09  747421.0
5    06  753265.0
4    05  798050.0
1    02  831150.0
0    01  858727.0

Historically, the best month for sales is December

# Finished!

Hope this was not too difficult and slicing and dicing the datasets was some fun. Now head on back to the course and provide the answers to the questions from this exercise.